### ISSUE TITLE
speed problem

### ISSUE THREAD

**Issue by gwlegion** [MEGAgameBoy/Sonic-Time-Twisted-PS-Vita-Port#1]
hi .... got some speed problems on a vite 1K ... got frame drops and zone with slow fps

**Comment by RinMaru** [comment]
Yea same here it is a bit laggy but not unplayable. It will slow down and stutter.

**Comment by rmsa5** [comment]
Indeed. Does the original feel also like it's more of a cruise vessel simulator than a classic Sonic game or is it that this port feels like it is running exclusively on the CPU ?

**Comment by RinMaru** [comment]
> Indeed. Does the original feel also like it's more of a cruise vessel simulator than a classic Sonic game or is it that this port feels like it is running exclusively on the CPU ?

probably is running in CPU mode because this is how the decomps play. they never got updated for VitaGL

**Comment by MDashK** [comment]
Use the overclock plugin on your Vita and overclock it to have 60 FPS most of the time.
The slowdowns are mostly due to the fact that the game is too heavy for the Vita, combined with the fact that the game is not using the async save method. It "freezes" temporarily to save the data.
Regarding the CPU usage: it's not the port/game itself. It's the GMS runner.
This is pretty much the best you can get playing this game on the Vita.

**Comment by RinMaru** [comment]
> Use the overclock plugin on your Vita and overclock it to have 60 FPS most of the time. The slowdowns are mostly due to the fact that the game is too heavy for the Vita, combined with the fact that the game is not using the async save method. It "freezes" temporarily to save the data. Regarding the CPU usage: it's not the port/game itself. It's the GMS runner. This is pretty much the best you can get playing this game on the Vita.

Its possible to optimize it better i gather? The engine could be reworked allot?

**Comment by MDashK** [comment]
> Its possible to optimize it better i gather? The engine could be reworked allot?

Sure, I guess, if you want to lose hundreds of hours optimizing it...
In all honesty, my personal opinion is that it's not worth that much effort.
The game is available on PC, Android, PS4, Xbox One/Series, Nintendo Switch...
If we are talking about portability, the Switch release is the better one and any laptop and/or Android phone/tablet nowadays runs this without any issue.
The Vita version is pretty much for those curious or those that really do not have anything else.

**Comment by gwlegion** [comment]
> that the game is too heavy for the Vita

are you serious ? vita can run gta san andreas without any slowdown, but your sonic in pixel art is too heavy ? 
Got so many exemple of very heavy games that run perfectly on vita ... 
Your code is bad, you don't want to lose time optimising it ... ok, but don't say it's the vita fault ...

**Comment by MDashK** [comment]
> are you serious ? vita can run gta san andreas without any slowdown, but your sonic in pixel art is too heavy ? Got so many exemple of very heavy games that run perfectly on vita ... Your code is bad, you don't want to lose time optimising it ... ok, but don't say it's the vita fault ...

... Are you serious? This comment of yours just shows how little knowledge you have on the matter.
Are you comparing games that run on specific engines with games that run on YYC runner?
GameMaker Studio runner is the issue. Not the game. It's not related if it uses 3D or pixel art.
It's a matter of code optimization. Not only on the game code, but on the runner it uses (aka engine).
There's a reason why GMS has 2 export ways: Normal and YoYo Compiled.
YYC is pretty much to try to optimize the code to have better performance. But it doesn't make miracles, specially on the Vita.
To sumarize: the game's code was aimed at PC and Android. As stated earlier, it would need hundreds of hours to optimize for the Vita, assuming the runner on the Vita would be able to run it full speed with such optimizations.
The runner, afaik, can't be optimized, since it's YoYo's own code. It's not even MY code.
So yeah, I never said Vita's fault. It's the lack of optimization in YoYo runner mainly.
There's a reason why I gave up on porting Relic Hunters Zero for the Vita.
But by all means: YOU grab the source code, and YOU optimize it, since you know so much on the matter and think it's so trivial.

**Comment by gwlegion** [comment]
have you tried to compile it for yoyo and use yoyo vita's implementation ?

**Comment by gwlegion** [comment]
you've said litteraly thae game is to heavy for the vita ... 
it's not right ... as you explained, it's matter to the engine that was not make for it and lock of optimisation . 
i can understant you don't got hundreds hours to optimise the engine ... but " the game is too heavy for the Vita" is not only a shortcut, it's just a lie . 
the game could prefectly work on vita, vita got enought power to run that kind of games a some way better, but your game depends to an engine that was not made for the vita and is very low on it ... 
The probleme don't come from vita's power but from the code optimisation (in our case, the engine code if i understand)

now i got a question : can GMS use SDL ? in fact, i really don't know anything about GMS ... 
after, i think you use kind of yoyo loader ?

**Comment by MDashK** [comment]
I said the game is too heavy for the Vita due to the lack of optimization in the GMS runner.
The engine IS GMS runner. It's the runner itself. It's where the game was created.
As I stated previously: it's more the lack of optimization on the runner, rather than the game itself.
Even with the game compiled with YYC, like it is, like I stated previously, it still would need a lot of work to make it 60 FPS permanent.
If you don't understand what you are talking about, don't go around calling others liars.
Learn about GMS, learn about the GMS Vita runner, then you can talk about it.
And if you have the boldness to state that you "really don't know anything about GMS ...", then you should have stayed quiet in the first place.

**Comment by MEGAgameBoy** [comment]
> you've said litteraly thae game is to heavy for the vita ... it's not right ... as you explained, it's matter to the engine that was not make for it and lock of optimisation . i can understant you don't got hundreds hours to optimise the engine ... but " the game is too heavy for the Vita" is not only a shortcut, it's just a lie . the game could prefectly work on vita, vita got enought power to run that kind of games a some way better, but your game depends to an engine that was not made for the vita and is very low on it ... The probleme don't come from vita's power but from the code optimisation (in our case, the engine code if i understand)
> 
> now i got a question : can GMS use SDL ? in fact, i really don't know anything about GMS ... after, i think you use kind of yoyo loader ?

Also take in consideration that we are 2 on the project and not a big triple A studio and we are making all of this stuff for free. So if you really want 60 fps just overclock the console for now.

**Comment by rmsa5** [comment]
@gwlegion you are the type of cancer user which breaks opensource initiatives. Let us see what are your programming credentials before you can be so salty with a couple of nice people enjoying porting stuff for the community in their personal time.

You suggest using SDL, but don't know a thing about SDL. Well STFU, and go port that game on your own.

This kind of stupid fuck is enraging.

@MEGAgameBoy  @MDashK thanks for the port. It's far from perfect. I hope clever tricks get found to make the experience more lightspeed-sonic-like in the future :) keep up the good work.

**Comment by gwlegion** [comment]
@mightysashiman  sorry, but when i read "The slowdowns are mostly due to the fact that the game is too heavy for the Vita" , i got angry ... i'm not a good dev, okay, but i know enought about hardware to know that vita is way enough powerfull to handle that type of games. 
i can understand that the engine was not designed for the vita at start, and the port is not optimal ... but the problem comes from the code , not from the hardware. 
I can also understand that the code optimisation could be very hard, in particular when it's not your code... 
But it still not a hardware problem. 
and yes i suggest using sdl ... maybe it's not pertinent, i'm not good enought to be sure ... i admit it. but some times, an idea, even a bad one could lead to a good one ... 

And for the "bad words" ... what can i answer ? thank you ? 

@MEGAgameBoy still good work ... hope it will be even better in the future. 
sorry about my "little enraging" ... i'm angry when someone says vita's hardware is not powerfull enough ... 
but i maintain my point of view : the problem is not that the vita is not powerfull enough . There is a code problem. But i unserstand the problem does not come from your code, and optimising someone else code is very hard and need a time you don't have ... 
maybe we could hope updates in the engine that make the port faster ...

**Comment by MDashK** [comment]
> @mightysashiman sorry, but when i read "The slowdowns are mostly due to the fact that the game is too heavy for the Vita" , i got angry ... i'm not a good dev, okay, but i know enought about hardware to know that vita is way enough powerfull to handle that type of games. i can understand that the engine was not designed for the vita at start, and the port is not optimal ... but the problem comes from the code , not from the hardware. I can also understand that the code optimisation could be very hard, in particular when it's not your code... But it still not a hardware problem. and yes i suggest using sdl ... maybe it's not pertinent, i'm not good enought to be sure ... i admit it. but some times, an idea, even a bad one could lead to a good one ...
> 
> And for the "bad words" ... what can i answer ? thank you ?
> 
> @MEGAgameBoy still good work ... hope it will be even better in the future. sorry about my "little enraging" ... i'm angry when someone says vita's hardware is not powerfull enough ... but i maintain my point of view : the problem is not that the vita is not powerfull enough . There is a code problem. But i unserstand the problem does not come from your code, and optimising someone else code is very hard and need a time you don't have ... maybe we could hope updates in the engine that make the port faster ...

You really don't understand English very well do you? If you did, you would have undersood the issue in the second reply I gave you.
I'm the first one to come at front and defend the power and magnificence of the Vita hardware.
IT'S. THE. RUNNER. If you STILL don't undestand what I mean when I state "runner", then you haven't undersood and learned anything yet.
Here's a good challenge for you: Since you love SDL so much: learn on how to implement SDL into GMS. But really LEARN.
Again: Before talking about any matter: Learn before everything else. I really hope you don't get this type of "enraged" behaviour dealing with real life issues.

**Comment by RinMaru** [comment]
I beat the game last night and while it was a bit sluggish i was able to 100% it. Thank you for your work and allowing me a convenient way of trying this fan game.

**Comment by MDashK** [comment]
**New version is up!**

**Comment by RinMaru** [comment]
> **New version is up!**

Lost my data lol it boots into menu faster and special stages start more smoothly so thank you. is it possible to use my PC save data on Vita?

**Comment by MEGAgameBoy** [comment]
> > **New version is up!**
> 
> Lost my data lol it boots into menu faster and special stages start more smoothly so thank you. is it possible to use my PC save data on Vita?

Yes it is, for that you need to take the save0.ini or save1.ini (in function of the number of saves you have) these files a located in "C:\Users\USERNAME\AppData\Local\SonicTimeTwisted", you need to transfer these file to "ur0:user/00/savedata/SOTT06722/" to your vita.
